src/compiler/checker.ts

```diff
@@ -22051,6 +22051,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // For example, `foo-${number}` is related to `foo-${string}` even though number isn't related to string.
                     instantiateType(source, reportUnreliableMapper);
                 }

```

Comment:
This is not specific to the PR, but how do we reach this point of the code in the example in the bug, when the return types are written without using a template literal type, e.g. `TypeMap[T0]`?
Also, this function in the new test:
```
function f3<T3 extends "a" | "b">(x: T3) {
    const test1: `${T3}` = x    
    const test2: T3 = "" as `${T3}`;    
}
```
Already works without those changes. So I guess I'm having a hard time figuring out how the bug arises and how changes in the PR are related to the bug, because it seems we already have some simplification going on for template literal types that look like `${T}`.

Reply:
> This is not specific to the PR, but how do we reach this point of the code in the example in the bug, when the return types are written without using a template literal type, e.g. TypeMap[T0]?

This is an interesting thing! And I don't really understand its details in depth 😅 

The source signature here goes through `instantiateSignatureInContextOf` and its result is being compared against the target. The template literal type gets inferred there through `inferToTemplateLiteralType` that gets invoked through `applyToParameterTypes`.

At that point, inference source and inference target are both template literal types. `inferTypesFromTemplateLiteralType` assesses that texts are the same and "converts" the types in the source using `getStringLikeTypeForType`. That wraps back `T` into a template literal type since it's not string-like.

So we end up with a template literal type as a single match returned by `inferTypesFromTemplateLiteralType` and the target (in `inferToTemplateLiteralType`) is still a template literal type as well. The match represents a placeholder match though and we loop through the types in the target, "unwrapping" `T` from the template literal type there and thus we end up inferring a template literal type (the placeholder match) into `T`.

---

> Already works without those changes. So I guess I'm having a hard time figuring out how the bug arises and how changes in the PR are related to the bug because it seems we already have some simplification going on for template literal types that look like ${T}.

Both of those assignments **don't** work right now: [TS playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMAzAHgCosQUwB5Q5gAmAzogEQCGFiAPpQEYUB8AFHgFyJYCUiAbwBQiUYggJSURISkBGbgAMAJAKwBfRYgC8iPGMQixEsFJk4pAJm5YdlWlXIq1KTQG4DQ9UA).

Reply:
Yeah, I mixed up my playgrounds again 😅
Ok, it seems to me like we're perhaps doing some unnecessary wrapping. Maybe we could check if a type parameter's constraint is string like in `getStringLikeTypeForType` and not wrap it in that case? If we did that, would we still need the changes in this PR?

Again, I'm not sure if I'm missing something and there's a reason why we do things the way they are right now, but I find it confusing.

Reply:
> If we did that, would we still need the changes in this PR?
 
I'm not entirely sure why the wrapping happens there but yes - if we'd remove it for some cases then it's likely that the issue at hand would be fixed. I considered that initially but I chose those other changes because they are also shipping what we have mentioned above:
```ts
function f3<T3 extends "a" | "b">(x: T3) {
    const test1: `${T3}` = x    
    const test2: T3 = "" as `${T3}`;    
}
```

So if you'd like me to explore that other route - that would "break" this test case.

Reply:
Ok, so I think this PR is probably good then. Maybe we should have both changes. In any case, I still have that question about whether ``` `${S}` ``` is the same as `S` when `S` is assignable to string...